lib/volcanic/authenticator/v1/privilege.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module Volcanic::Authenticator
+  module V1
+    # Privilege api
+    class Privilege < Common
+      class << self
+        def path
+          'api/v1/subject'
+        end
+
+        def exception
+          PrivilegeError
+        end
+
+        # Filtering by service, a single hash is returned.
+        def find_by_service(subject, service)
+          privilege = perform_get_and_parse(
+            exception,
+            "#{path}/privileges?filter[subject]=#{subject}&filter[service]=#{service}"
+          )
+          return nil if privilege.blank?
+
+          new(privilege.transform_keys!(&:to_sym))
+        end
+      end
+
+      def initialize(**opts)
+        @service = opts.fetch(:service)
+        @permissions = opts.fetch(:permissions)
+      end
+
+      def privileges
+        @privileges ||= permissions.map do |perm|
+          perm.fetch('privileges', []).map do |priv|
+            {
+              name: perm.fetch('name'),
+              scope: priv.fetch('scope'),
+              allow: priv.fetch('allow')
+            }
+          end
+        end.flatten
+      end
+
+      attr_accessor :service, :permissions
+    end
+  end
+end
```

Comment:
ok. clearly there is something amiss here

Let's think about a simple interface.

A privilege contains the following data items:

```ruby
class Privilege
  attr_accessor :id, :permission_id, :scope, :allow, :permission_group_id
end
```

in that, `permission` or `permission_group` may be nil, but at least one must be present.

A `permission_group` is a link allowing many permissions to be applied together to the same scope and allow.

This means that we could extend the interface to be:
```ruby
class Privilege
  attr_accessor :id, :permission_id, :scope, :allow, :permission_group_id

  def permission_ids
    [permission_id, PermissionGroup.new(permission_group_id).permission_ids].flatten.compact
  end
end
```

to be more useful, we could extend further to be:
```ruby
class Privilege
  attr_accessor :id, :permission_id, :scope, :allow, :permission_group_id

  def permission_ids
    [permission_id, PermissionGroup.find_by_id(permission_group_id).permission_ids].flatten.compact
  end

  def permissions
    permission_ids.map { |perm_id| Permission.find_by_id(perm_id) }
  end
end
```
For some responses (like the GET subject/privileges), we may be able to generate the linked permissions without having to fetch them as well, then handle that with memoisation.

Note that this does not include `service`. The service is related to the permissions, not the privilege, so a privilege *may* be transitively related to multiple services.

Starting with the basic data on a privilege object, we should be able to create appropriate accessors that pull that data from the API and can create collections of instances based off that.

To further simplify, we can accept that the initial usecase is just to get the privileges for a single subject. This means that we can ignore permission groups (the privileges returned from that API are de-normalised already).

Reply:
Agreed on the principals here - however the results returned from the API do not mention permission groups:
```
{
    "requestID": "offline_awsRequestId_5364920558730266",
    "response": {
        "service": "auth",
        "permissions": [
            {
                "id": 1,
                "name": "identity:create",
                "privileges": [
                    {
                        "id": 1,
                        "scope": "vrn:local:-1:identity/*",
                        "allow": true
                    }
                ]
            },
...
```

Maybe discuss in person tomorrow 👍

Reply:
tl; dr: there are 2 different APIs. The one you're using is the subject/privileges API, NOT the privilege API. The subject/privileges API is slightly different in that it returns you de-normalised privileges to make it easier for the client to process for auth. The privileges API returns a very different format (see the GET privileges example in Postman) with the focus being on the privileges themselves rather than the privileges by permission by service.